cranelift/codegen/src/legalizer/mod.rs

```diff
@@ -157,160 +157,56 @@ pub fn simple_legalize(func: &mut ir::Function, cfg: &mut ControlFlowGraph, isa:
-                }
-                InstructionData::BinaryImm64 {
-                    opcode: ir::Opcode::SremImm,
-                    arg,
-                    imm,
-                } => {
-                    let ty = pos.func.dfg.value_type(arg);
-                    let imm = pos.ins().iconst(ty, imm);
-                    pos.func.dfg.replace(inst).srem(arg, imm);
-                }
-                InstructionData::BinaryImm64 {
-                    opcode: ir::Opcode::UdivImm,
-                    arg,
-                    imm,
-                } => {
+                InstructionData::BinaryImm64 { opcode, arg, imm } => {
                     let ty = pos.func.dfg.value_type(arg);
-                    let imm = pos.ins().iconst(ty, imm);
-                    pos.func.dfg.replace(inst).udiv(arg, imm);
-                }
-                InstructionData::BinaryImm64 {
-                    opcode: ir::Opcode::UremImm,
-                    arg,
-                    imm,
-                } => {
-                    let ty = pos.func.dfg.value_type(arg);
-                    let imm = pos.ins().iconst(ty, imm);
-                    pos.func.dfg.replace(inst).urem(arg, imm);
+                    let imm = if ty == I128 {
+                        let imm = pos.ins().iconst(I64, imm);
+                        pos.ins().sextend(I128, imm)
+                    } else {
+                        pos.ins().iconst(ty.lane_type(), imm)
+                    };
+
+                    let replace = pos.func.dfg.replace(inst);
+                    match opcode {
+                        // bitops
+                        ir::Opcode::BandImm => replace.band(arg, imm),
+                        ir::Opcode::BorImm => replace.bor(arg, imm),
+                        ir::Opcode::BxorImm => replace.bxor(arg, imm),
+                        // bitshifting
+                        ir::Opcode::IshlImm => replace.ishl(arg, imm),
+                        ir::Opcode::RotlImm => replace.rotl(arg, imm),
+                        ir::Opcode::RotrImm => replace.rotr(arg, imm),
+                        ir::Opcode::SshrImm => replace.sshr(arg, imm),
+                        ir::Opcode::UshrImm => replace.ushr(arg, imm),
+                        // math
+                        ir::Opcode::IaddImm => replace.iadd(arg, imm),
+                        ir::Opcode::IrsubImm => replace.isub(imm, arg), // note: arg order reversed
+                        ir::Opcode::ImulImm => replace.imul(arg, imm),
+                        ir::Opcode::SdivImm => replace.sdiv(arg, imm),
+                        ir::Opcode::SremImm => replace.srem(arg, imm),
+                        ir::Opcode::UdivImm => replace.udiv(arg, imm),
+                        ir::Opcode::UremImm => replace.urem(arg, imm),
+                        // comparisons
+                        ir::Opcode::IfcmpImm => replace.ifcmp(arg, imm),
+                        _ => unimplemented!(),
```

Comment:
It looks like it's a behavioral change if we panic here. The outer match just skips the instruction if it doesn't have an expansion for it. Maybe that's why the previous implementation duplicated so much code? I think @cfallin probably needs to weigh in on how this case should be handled.

Gotta say, though, I _really_ like how much shorter your version is. I hope we can preserve that clarity even if this bit has to change.

It might help to introduce a function that's something like this (but I haven't tested this code, let alone compiled it):

```rust
fn imm_const(pos: &mut FuncCursor, arg: Value, imm: Imm64) -> Value {
    let ty = pos.func.dfg.value_type(arg);
    let imm = pos.ins().iconst(ty, imm);
    if ty == I128 {
        let imm = pos.ins().iconst(I64, imm);
        pos.ins().sextend(I128, imm)
    } else {
        pos.ins().iconst(ty.lane_type(), imm)
    }
}
```

Reply:
Yes, we don't want to hit an `unimplemented!()` here; I think we just want an empty match-arm body (`_ => {}`) instead. That way we still construct the `imm` but we just don't do anything with it.

Reply:
Wouldn't an empty match-arm body there cause `simple_legalize` to loop, trying to legalize the same instruction repeatedly?

Reply:
The toplevel loop is a `while let Some(inst) = pos.next_inst()`, so it's stepping through insts with no action taken in the loop body. Though actually the more precise thing to do is to replicate the fallback `_ => { ... }` below, which sets `prev_pos` first then continues, I think.

Reply:
The statement after the `match` is `pos.set_position(prev_pos);` and the intent appears to be to re-examine the result of every legalization. So it looks to me like termination of this function relies on every match arm replacing the instruction that it originally matched on. But yes, copying the `prev_pos` assignment from the fallback case should work, I think.